.github/workflows/unit_tests.yml

```diff
@@ -0,0 +1,70 @@
+name: Unit tests
+
+on:
+  pull_request:
+    branches:
+      - main
+
+concurrency:
+  group: ${{ github.ref }}-${{ github.workflow }}
+  cancel-in-progress: true
+
+env:
+  CARGO_TERM_COLOR: always
+  CARGO_INCREMENTAL: 0
+
+jobs:
+  checks:
+    name: Run clippy and unit tests
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout source code
+        uses: actions/checkout@v3
+
+      - name: Run cargo fmt
+        uses: actions-rs/cargo@v1
+        with:
+          command: fmt
+          args: --all
+
+      - name: Run clippy
+        uses: actions-rs/cargo@v1
+        with:
+          command: clippy
+          args: --all-targets -- --no-deps -D warnings
+
+      - name: Run unit tests
+        uses: actions-rs/cargo@v1
+        with:
+          command: test
+
+  build:
+    name: Build the contract
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout source code
+        uses: actions/checkout@v3
+
+      - name: Cache Crates
+        uses: actions/cache@v3
+        with:
+          path: ~/.cargo
+          key: ${{ runner.os }}-contract-build-env
```

Comment:
This will  cache `~/.cargo`  across different Rust toolchain versions. Is that intended?

Reply:
Yes, it is

Reply:
but why? We probably won't need more than 1 version at the time so why cache them all?

Reply:
Hmm, here we are doing everything with the default stable Rust, without resorting to "toolchain versioning" we do in other repos. I'm not sure, but I think that if the `stable` toolchain gets updated "behind the curtains", this workflow will just keep working without bloating the cache...

Reply:
About this cache: with GH runners it is fine to choose whatever caching mechanism as it is separated from self-hosted machines we use anyway. So as long as it improves something, we can use it. 

We had some issues in cargo cache last time, but we did not get to the bottom of the problem, so maybe it's worth to give it a second go now. I'm just curious for example, does this influence `cargo contract install`? IT should right? But the time of the build is always 19 minutes.

Reply:
Yes, caching is here precisely for avoiding loooong cargo contract compilation time. And it seems to work:
https://github.com/Cardinal-Cryptography/PSP22/actions/runs/6810159396/job/18517919901